doc/harry.man
.\" Automatically generated by Pod::Man 2.27 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "HARRY 1"
.TH HARRY 1 "2016-12-23" "Harry 0.4.3" "User Manual"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
harry \- A tool for measuring string similarity
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBharry\fR [\fBoptions\fR] [\fB\-c\fR \fIconfig\fR] \fIinput\fR [\fIinput\fR] \fIoutput\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBharry\fR is a small tool for measuring the similarity of strings. The tool
supports common distance and kernel functions for strings as well as some
exotic similarity measures.  The focus of \fBharry\fR lies on implicit
similarity measures, that is, comparison functions that do not give rise to
an explicit vector space.  Examples of such similarity measures are the
Levenshtein distance, the Jaro-Winkler distance or the normalized
compression distance.
.PP
During operation \fBharry\fR loads a set of strings from \fIinput\fR, computes the
specified similarity measure and writes a matrix of similarity values to
\&\fIoutput\fR.  If one \fIinput\fR is given, \fBharry\fR computes the similarities
between all strings in \fIinput\fR.  If two \fIinput\fR sources are provided,
\&\fBharry\fR computes only the similarities between the two sources.  The
similarity measure can be computed based on the granularity of bits, bytes
as well as tokens contained in the strings.  The configuration of this
process, such as the input format, the similarity measure and the output
format, are specified in the file \fIconfig\fR and can be additionally refined
using command-line options.
.PP
.Vb 7
\&    .\-\-\-\-\-\-\-\-\-.                        .\-\-\-\-\-\-\-\-\-\-.
\&    |  lines  |                        |   text   |
\&    |   dir   |   \e   .\-\-\-\-\-\-\-\-\-.   /  |  libsvm  |
\&    |  stdin  |   \-\-  |  Harry  |  \-\-  |  matlab  |
\&    |   ...   |   /   \*(Aq\-\-\-\-\-\-\-\-\-\*(Aq   \e  |   ...    |
\&    \*(Aq\-\-\-\-\-\-\-\-\-\*(Aq                        \*(Aq\-\-\-\-\-\-\-\-\-\-\*(Aq
\&       Input      Similarity Measures     Output
.Ve
.PP
\&\fBharry\fR is implemented using OpenMP, such that the computation time for a
set of strings scales linear with the number of available \s-1CPU\s0 cores.
Moreover, efficient implementations of several similarity measures,
effective caching of similarity values and low-overhead locking further
speedup the computation.
.PP
\&\fBharry\fR complements the tool \fBsally\fR(1) that embeds strings in a vector
space and allows computing vectorial similarity measures, such as the cosine
distance and the bag-of-words kernel.
.SH "CONFIGURATION"
.IX Header "CONFIGURATION"
The configuration of \fBharry\fR is provided by a configuration file.  This
file is structured into the three sections \fBinput\fR, \fBmeasures\fR and
\&\fBoutput\fR, which define the parameters of the input format, the similarity
measures and the output format, respectively.  If no configuration file is
provided, \fBharry\fR resorts to a default configuration.  This default
configuration can be printed using the command-line option \fB\-D\fR (see
\&\fB\s-1OPTIONS\s0\fR).
.SS "Input formats"
.IX Subsection "Input formats"
\&\fBharry\fR supports different formats for reading sets of strings, which may
range from plain files to directories and other structured resources.  The
input format is specified in the configuration file of \fBharry\fR, but can
also be defined on the command line using the option \fB\-i\fR (see \fB\s-1OPTIONS\s0\fR).
Following is a list of supported input formats:
.IP "\fBinput = {\fR" 4
.IX Item "input = {"
.RS 4
.PD 0
.ie n .IP "\fBinput_format = ""lines"";\fR" 4
.el .IP "\fBinput_format = ``lines'';\fR" 4
.IX Item "input_format = lines;"
.PD
This parameter specifies the input format.
.RS 4
.ie n .IP "\fI""lines""\fR" 14
.el .IP "\fI``lines''\fR" 14
.IX Item "lines"
The input strings are available as lines in a text file. The name of the
file is given as \fIinput\fR to \fBharry\fR.  The lines need to be separated by
newline and may not contain the \s-1NUL\s0 character.  Labels can be extracted from
each line using a regular expression (see \fBlines_regex\fR).
.ie n .IP "\fI""stdin""\fR" 14
.el .IP "\fI``stdin''\fR" 14
.IX Item "stdin"
The input strings are provided on standard input (stdin) as text lines.  The
lines need to be separated by newline and may not contain the \s-1NUL\s0 character. 
Labels can be extracted from each line using a regular expression (see
\&\fBlines_regex\fR).  This input format is also enabled when \fIinput\fR is set to
\&\fI\-\fR, otherwise \fIinput\fR is ignored.
.ie n .IP "\fI""dir""\fR" 14
.el .IP "\fI``dir''\fR" 14
.IX Item "dir"
The input strings are available as binary files in a directory and the
name of the directory is given as \fIinput\fR to \fBharry\fR. The suffixes
of the files are used as labels for the strings.
.ie n .IP "\fI""arc""\fR" 14
.el .IP "\fI``arc''\fR" 14
.IX Item "arc"
The input strings are available as binary files in a compressed archive,
such as a zip or tgz archive.  The name of the archive is given as \fIinput\fR
to \fBharry\fR.  The suffixes of the files are used as labels for the strings.
.ie n .IP "\fI""fasta""\fR" 14
.el .IP "\fI``fasta''\fR" 14
.IX Item "fasta"
The input strings are available in \s-1FASTA\s0 format. The name of the file is
given as \fIinput\fR to \fBharry\fR.  Labels can be extracted from the description
of each sequence using a regular expression (see \fBfasta_regex\fR).  Comments
are allowed if they are preceded by either ';' or '>'.
.ie n .IP "\fI""raw""\fR" 14
.el .IP "\fI``raw''\fR" 14
.IX Item "raw"
The input strings are provided on standard input (stdin) in \*(L"raw\*(R" format.
This input module is designed to efficiently interface with other
environments.  The binary format for strings has the form
.Sp
.Vb 3
\&  | len (uint32) | array (uint8) ... |
\&  | len (uint32) | array (uint8) ... |
\&  | ...
.Ve
.Sp
where \fIlen\fR is a 32\-bit unsigned integer in host byte order indicating the
length of the following \fIarray\fR containing the string data in bytes. 
Labels cannot be extracted from this representation.  This input format is
also enabled when \fIinput\fR is set to \fI=\fR, otherwise \fIinput\fR is ignored.
.RE
.RS 4
.RE
.IP "\fBchunk_size = 256;\fR" 4
.IX Item "chunk_size = 256;"
To enable an efficient processing of large data sets, \fBharry\fR loads strings
in chunks.  This parameter defines the number of strings in one of these
chunks.  Depending on the lengths and type of the strings, this parameter
can be adjusted to improve loading times.
.IP "\fBdecode_str = false;\fR" 4
.IX Item "decode_str = false;"
If this parameter is set to \fItrue\fR, \fBharry\fR automatically decodes strings
that contain URI-encoded characters.  That is, substrings of the form \f(CW%XX\fR
are replaced with the byte corresponding to the hexadecimal number \s-1XX.\s0
.ie n .IP "\fBfasta_regex = "" (\e\e+|\-)?[0\-9]+"";\fR" 4
.el .IP "\fBfasta_regex = `` (\e\e+|\-)?[0\-9]+'';\fR" 4
.IX Item "fasta_regex = (+|-)?[0-9]+;"
The \s-1FASTA\s0 format allows to equip each string with a short description.  In
several data sets this description contains a numerical label which can be
used for supervised learning tasks.  The parameter defines a regular
expression that matches these numerical labels, such as +1 and \-1.
.ie n .IP "\fBlines_regex = ""^(\e\e+|\-)?[0\-9]+"";\fR" 4
.el .IP "\fBlines_regex = ``^(\e\e+|\-)?[0\-9]+'';\fR" 4
.IX Item "lines_regex = ^(+|-)?[0-9]+;"
If the strings are available as text lines, the parameter can be used to
extract a numerical label from the strings.  The parameter is a regular
expression matching labels, such as +1 and \-1.
.IP "\fBreverse_str = false;\fR" 4
.IX Item "reverse_str = false;"
If this parameter is set to \fItrue\fR, the characters of all input strings
will be reversed.  Such reversing might help in situations where the reading
direction of the input strings is unspecified.
.ie n .IP "\fBstoptoken_file = """";\fR" 4
.el .IP "\fBstoptoken_file = ``'';\fR" 4
.IX Item "stoptoken_file = """";"
Stop tokens (irrelevant tokens) can be filtered from the strings by providing
a file containing these tokens; one per line.  Non-printable characters can
be escaped using \s-1URI\s0 encoding (%XX).  Stop tokens can only be filtered, if
the \fBgranularity\fR is set to \fItokens\fR.
.IP "\fBsoundex = false;\fR" 4
.IX Item "soundex = false;"
All tokens in the strings are mapped to the soundex index. For example,
\&\*(L"Pfister\*(R" is mapped to \*(L"P236\*(R" and \*(L"Jackson\*(R" to \*(L"J250\*(R".  The soundex index
has been originally designed for comparing names, however, in \fBharry\fR it
can be applied to all sorts of tokens, if they are composed of alphabetic
letters.  Punctation characters are ignored and thus the string \*(L"Hey, I am
here with Harry!\*(R", gets mapped to \*(L"H000 I000 A500 H600 W300 H600\*(R".
.RE
.RS 4
.RE
.IP "\fB};\fR" 4
.IX Item "};"
.SS "Similarity Measures"
.IX Subsection "Similarity Measures"

src/harry.c
/*
 * Harry - A Tool for Measuring String Similarity
 * Copyright (C) 2013-2014 Konrad Rieck (konrad@mlsec.org)
 * --
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 3 of the License, or (at your
 * option) any later version.  This program is distributed without any
 * warranty. See the GNU General Public License for more details.
 */

#include "config.h"
#include "common.h"
#include "harry.h"
#include "hconfig.h"
#include "util.h"
#include "input.h"
#include "measures.h"
#include "output.h"
#include "vcache.h"
#include "hmatrix.h"

/* Global variables */
int verbose = 0;
int log_line = 0;
config_t cfg;

static int print_conf = 0;
static char *measure = NULL;
static int benchmark = 0;

/* Option string */
#define OPTSTRING       "p:n:g:a:m:c:i:o:d:x:y:zs:ulvqGVhMCD"

/**
 * Array of options of getopt_long()
 */
static struct option longopts[] = {
    {"input_format", 1, NULL, 'i'},
    {"decode_str", 0, NULL, 1000},
    {"reverse_str", 0, NULL, 1001},
    {"stoptoken_file", 1, NULL, 1002},
    {"soundex", 0, NULL, 1003},
    {"benchmark", 1, NULL, 1004},
    {"save_indices", 0, NULL, 1005},
    {"save_labels", 0, NULL, 1006},
    {"save_sources", 0, NULL, 1007},
    {"output_format", 1, NULL, 'o'},
    {"compress", 0, NULL, 'z'},
    {"measure", 1, NULL, 'm'},
    {"token_delim", 1, NULL, 'd'},
    {"num_threads", 1, NULL, 'n'},
    {"cache_size", 1, NULL, 'a'},
    {"granularity", 1, NULL, 'g'},
    {"global_cache", 0, NULL, 'G'},
    {"config_file", 1, NULL, 'c'},
    {"verbose", 0, NULL, 'v'},
    {"log_line", 0, NULL, 'l'},
    {"precision", 1, NULL, 'p'},
    {"print_measures", 0, NULL, 'M'},
    {"print_config", 0, NULL, 'C'},
    {"print_defaults", 0, NULL, 'D'},
    {"quiet", 0, NULL, 'q'},
    {"version", 0, NULL, 'V'},
    {"help", 0, NULL, 'h'},
    {"x_range", 1, NULL, 'x'},
    {"y_range", 1, NULL, 'y'},
    {"split", 1, NULL, 's'},
    {NULL, 0, NULL, 0}
};

/**
 * Prints version and copyright information to a file stream
 * @param f File pointer
 * @param p Prefix character
 * @param m Message
 * @return number of written characters
 */
int harry_version(FILE *f, char *p, char *m)
{
    return fprintf(f, "%sHarry %s - %s\n", p, PACKAGE_VERSION, m);
}

/**
 * Prints version and copyright information to a file stream
 * @param z File pointer
 * @param p Prefix character
 * @param m Message
 * @return number of written characters
 */
int harry_zversion(gzFile z, char *p, char *m)
{
    return gzprintf(z, "%sHarry %s - %s\n", p, PACKAGE_VERSION, m);
}

/**
 * Print configuration
 * @param msg Text to add to output
 */
static void print_config(char *msg)
{
    harry_version(stdout, "# ", msg);
    config_print(&cfg);
}

/**
 * Print usage of command line tool
 */
static void print_usage(void)
{
    printf("Usage: harry [options] <input> [<input>] <output>\n"
           "\nI/O options\n"
           "  -i,  --input_format <format>    Set input format for strings.\n"
           "       --decode_str               Enable URI-decoding of strings.\n"
           "       --reverse_str              Reverse (flip) all strings.\n"
           "       --stoptoken_file <file>    Provide a file with stop tokens.\n"
           "       --soundex                  Enable soundex encoding of tokens.\n"
           "       --benchmark <seconds>      Perform benchmark run.\n"
           "  -o,  --output_format <format>   Set output format for matrix.\n"
           "  -p,  --precision <num>          Set precision of output.\n"
           "  -z,  --compress                 Enable zlib compression of output.\n"
           "       --save_indices             Save indices of strings.\n"
           "       --save_labels              Save labels of strings.\n"
           "       --save_sources             Save sources of strings.\n"
           "\nModule options:\n"
           "  -m,  --measure <name>           Set similarity measure.\n"
           "  -g,  --granularity <type>       Set granularity: bytes, bits, tokens.\n"
           "  -d,  --token_delim <delim>      Set delimiters for tokens.\n"
           "  -n,  --num_threads <num>        Set number of threads.\n"
           "  -a,  --cache_size <size>        Set size of cache in megabytes.\n"
           "  -G,  --global_cache             Enable global cache.\n"
           "  -x,  --x_range <start>:<end>    Set the index range (x) of strings.\n"
           "  -y,  --y_range <start>:<end>    Set the index range (y) of strings.\n"
           "  -s,  --split <blocks>:<idx>     Split matrix into blocks and compute one.\n"
           "\nGeneric options:\n"
           "  -c,  --config_file <file>       Set configuration file.\n"
           "  -v,  --verbose                  Increase verbosity.\n"
           "  -l,  --log_line                 Print a log line every minute\n"
           "  -q,  --quiet                    Be quiet during processing.\n"
           "  -M,  --print_measures           Print list of similarity measures\n"
           "  -C,  --print_config             Print the current configuration.\n"
           "  -D,  --print_defaults           Print the default configuration.\n"
           "  -V,  --version                  Print version and copyright.\n"
           "  -h,  --help                     Print this help screen.\n"
           "\n");
}

/**
 * Print version of Harry tool
 */
static void print_version(void)
{
    printf("Harry %s - A Tool for Measuring String Similarity\n"
           "Copyright (c) 2013-2014 Konrad Rieck (konrad@mlsec.org)\n",
           PACKAGE_VERSION);
}

/**
 * Parse command line options
 * @param argc Number of arguments
 * @param argv Argument values
 * @param in Return pointer to input filename
 * @param in Return pointer to optional input filename
 * @param out Return pointer to output filename
 */
static void harry_parse_options(int argc, char **argv, char **in,
                                char **in2, char **out)
{
    int ch;
    const char *str;
    optind = 0;

    while ((ch = getopt_long(argc, argv, OPTSTRING, longopts, NULL)) != -1) {
        switch (ch) {
        case 'c':
            /* Skip. See harry_load_config(). */
            break;
        case 'i':
            config_set_string(&cfg, "input.input_format", optarg);
            break;
        case 1000:
            config_set_bool(&cfg, "input.decode_str", CONFIG_TRUE);
            break;
        case 1001:
            config_set_bool(&cfg, "input.reverse_str", CONFIG_TRUE);
            break;
        case 1002:
            config_set_string(&cfg, "input.stoptoken_file", optarg);
            break;
        case 1003:
            config_set_bool(&cfg, "input.soundex", CONFIG_TRUE);
            break;
        case 1004:
            benchmark = atoi(optarg);
            break;
        case 1005:
            config_set_bool(&cfg, "output.save_indices", CONFIG_TRUE);
            break;
        case 1006:
            config_set_bool(&cfg, "output.save_labels", CONFIG_TRUE);
            break;
        case 1007:
            config_set_bool(&cfg, "output.save_sources", CONFIG_TRUE);
            break;
        case 'o':
            config_set_string(&cfg, "output.output_format", optarg);
            break;
        case 'p':
            config_set_int(&cfg, "output.precision", atoi(optarg));
            break;
        case 'm':
            config_set_string(&cfg, "measures.measure", optarg);
            break;
        case 'd':
            config_set_string(&cfg, "measures.token_delim", optarg);
            /* Compatibility warning */
            config_lookup_string(&cfg, "measures.granularity", &str);
            if (strcasecmp(str, "tokens") && strlen(optarg) > 0)
                warning("Granularity is %s. Delimiters are ignored.", str);
            break;
        case 'n':
            config_set_int(&cfg, "measures.num_threads", atoi(optarg));
            break;
        case 'a':
            config_set_int(&cfg, "measures.cache_size", atoi(optarg));
            break;
        case 'G':
            config_set_bool(&cfg, "measures.global_cache", CONFIG_TRUE);
            break;
        case 'g':
            config_set_string(&cfg, "measures.granularity", optarg);
            break;
        case 'z':
            config_set_bool(&cfg, "output.compress", CONFIG_TRUE);
            break;
        case 'x':
            config_set_string(&cfg, "measures.x_range", optarg);
            break;
        case 'y':
            config_set_string(&cfg, "measures.y_range", optarg);
            break;
        case 's':
            config_set_string(&cfg, "measures.split", optarg);
            break;
        case 'q':
            verbose = 0;
            log_line = 0;
            break;
        case 'l':
            log_line = 1;
            break;
        case 'v':
            verbose++;
            break;
        case 'M':
            measure_print();
            exit(EXIT_SUCCESS);
            break;
        case 'D':
            print_config("Default configuration");
            exit(EXIT_SUCCESS);
            break;
        case 'C':
            print_conf = 1;
            break;
        case 'V':
            print_version();
            exit(EXIT_SUCCESS);
            break;
        case 'h':
        case '?':
            print_usage();
            exit(EXIT_SUCCESS);
            break;
        }
    }

    /* Check configuration */
    if (!config_check(&cfg)) {
        exit(EXIT_FAILURE);
    }

    /* We are through with parsing. Print the config if requested */
    if (print_conf) {
        print_config("Current configuration");
        exit(EXIT_SUCCESS);
    }

    argc -= optind;
    argv += optind;

    /* Check for input and output arguments */
    if (argc == 2) {
        *in = argv[0];
        *out = argv[1];
    } else if (argc == 3) {
        *in = argv[0];
        *in2 = argv[1];
        *out = argv[2];
    } else {
        print_usage();
        exit(EXIT_FAILURE);
    }

    /* Check for stdin and stdout "filenames" */
    if (!strcmp(*in, "-"))
        config_set_string(&cfg, "input.input_format", "stdin");
    if (!strcmp(*in, "="))
        config_set_string(&cfg, "input.input_format", "raw");
    if (!strcmp(*out, "-"))
        config_set_string(&cfg, "output.output_format", "stdout");
    if (!strcmp(*out, "="))
        config_set_string(&cfg, "output.output_format", "raw");

}


/**
 * Load the configuration of Harry
 * @param argc number of arguments
 * @param argv arguments
 */
static void harry_load_config(int argc, char **argv)
{
    char *cfg_file = NULL;
    int ch;

    /* Check for config file in command line */
    while ((ch = getopt_long(argc, argv, OPTSTRING, longopts, NULL)) != -1) {
        switch (ch) {
        case 'c':
            cfg_file = optarg;
            break;
        case '?':
            print_usage();
            exit(EXIT_SUCCESS);
            break;
        default:
            /* empty */
            break;
        }
    }

    /* Init and load configuration */
    config_init(&cfg);

    if (cfg_file != NULL) {
        if (config_read_file(&cfg, cfg_file) != CONFIG_TRUE)
            fatal("Could not read configuration (%s in line %d)",
                  config_error_text(&cfg), config_error_line(&cfg));
    }

    /* Check configuration and set defaults */
    if (!config_check(&cfg)) {
        exit(EXIT_FAILURE);
    }

}

/**
 * Init the Harry tool
 * @param argc number of arguments
 * @param argv arguments
 */
static void harry_init()
{
    const char *cfg_str;
    cfg_int nthreads = 0;

    if (verbose > 1)
        config_print(&cfg);

    /* Init value cache */
    vcache_init();

    /* Configure module (init as first) */
    config_lookup_string(&cfg, "measures.measure", &cfg_str);
    measure = measure_config(cfg_str);

    config_lookup_int(&cfg, "measures.num_threads", &nthreads);
#ifdef HAVE_OPENMP
    if (nthreads <= 0)
        nthreads = omp_get_num_procs();
    omp_set_num_threads(nthreads);
#else
    warning("Harry has been compiled without OpenMP support.");
#endif

    /* Load stop tokens */
    config_lookup_string(&cfg, "input.stoptoken_file", &cfg_str);
    if (strlen(cfg_str) > 0)
        stoptokens_load(cfg_str);
}

/**
 * Read a set of strings to memory from input
 * @param input Input filename
 * @param input2 Optional input filename
 * @param num Pointer to number of strings
 * @return array of string objects
 */
static hstring_t *harry_read(char *input, char *input2, int *num)
{
    const char *cfg_str;
    int i, read;
    cfg_int chunk;
    hstring_t *strs = NULL;
    char buf[128];

    /* Get chunk size */
    config_lookup_int(&cfg, "input.chunk_size", &chunk);

    /* Open input */
    config_lookup_string(&cfg, "input.input_format", &cfg_str);
    info_msg(1, "Opening input '%0.40s' [%s].", input, cfg_str);
    input_config(cfg_str);
    if (!input_open(input))
        fatal("Could not open input source");

    info_msg(1, "Reading strings from %s", input);
    for (*num = 0, read = chunk; read == chunk; *num += read) {
        /* Allocate memory for strings */
        strs = realloc(strs, (*num + chunk) * sizeof(hstring_t));
        if (!strs)
            fatal("Could not allocate memory for strings");

        /* Read chunk */
        read = input_read(strs + *num, chunk);
    }

    /* Close input */
    input_close();

    /* Second input available */
    if (input2) {
        /* Store length of first input */
        int len1 = *num;
        if (!input_open(input2))
            fatal("Could not open second input source");

        info_msg(1, "Reading strings from %s", input2);
        for (read = chunk; read == chunk; *num += read) {
            /* Allocate memory for strings */
            strs = realloc(strs, (*num + chunk) * sizeof(hstring_t));
            if (!strs)
                fatal("Could not allocate memory for strings");

            /* Read chunk */
            read = input_read(strs + *num, chunk);
        }

        /* Close input */
        input_close();

        /* Overwrite x-range and y-range */
        snprintf(buf, 128, "%d:%d", 0, len1);
        config_set_string(&cfg, "measures.x_range", buf);
        snprintf(buf, 128, "%d:%d", len1, *num);
        config_set_string(&cfg, "measures.y_range", buf);
    }

    /* Symbolize strings if requested */
    for (i = 0; i < *num; i++)
        strs[i] = hstring_preproc(strs[i]);

    return strs;
}

/**
 * Compare a set of string objects
 * @param strs Array of string objects
 * @param num Number of strings
 * @return Matrix of similarity valurs
 */
void harry_compute(hmatrix_t *mat, hstring_t *strs, int num)
{
    /* Compute matrix */
#ifdef HAVE_OPENMP
    info_msg(1, "Computing similarity measure '%s' with %d threads.",
             measure, omp_get_max_threads());
#else
    info_msg(1, "Computing similarity measure '%s'", measure);
#endif
    hmatrix_compute(mat, strs, measure_compare);
}


/**
 * Benchmark runtime
 * @param strs Array of string objects
 * @param num Number of strings
 * @return Matrix of similarity valurs
 */
void harry_benchmark(hmatrix_t *mat, hstring_t *strs, int num)
{
    /* Compute matrix */
#ifdef HAVE_OPENMP
    info_msg(1, "Benchmarking similarity measure '%s' (%d thrd; %d sec).",
             measure, omp_get_max_threads(), benchmark);
    float cmps = hmatrix_benchmark(mat, strs, measure_compare, benchmark);
    printf("%.0f comparisons; %d seconds; %d threads;\n", cmps, benchmark,
           omp_get_max_threads());
#else
    info_msg(1, "Benchmarking similarity measure '%s' (%d sec).",
             measure, benchmark);
    float cmps = hmatrix_benchmark(mat, strs, measure_compare, benchmark);
    printf("%.0f comparisons; %d seconds;\n", cmps, benchmark);
#endif
}

/**
 * Init and allocate matrix for computation
 * @param strs Array of string objects
 * @param num Number of strings
 * @return Empty matrix
 */
static hmatrix_t *harry_alloc(hstring_t *strs, int num)
{
    char *cfg_str;
    int i;

    hmatrix_t *mat = hmatrix_init(strs, num);

    /* Set ranges */
    config_lookup_string(&cfg, "measures.x_range", (const char **) &cfg_str);
    hmatrix_xrange(mat, cfg_str);
    config_lookup_string(&cfg, "measures.y_range", (const char **) &cfg_str);
    hmatrix_yrange(mat, cfg_str);

    /* Set matrix split */
    config_lookup_string(&cfg, "measures.split", (const char **) &cfg_str);
    hmatrix_split(mat, cfg_str);

    /* Free unused memory */
    for (i = 0; i < num; i++) {
        if (i < mat->x.i && i < mat->y.i)
            hstring_destroy(&strs[i]);
        if (i >= mat->x.n && i >= mat->y.n)
            hstring_destroy(&strs[i]);
    }

    /* Allocate matrix */
    if (!hmatrix_alloc(mat))
        fatal("Could not allocate matrix for similarity measure");

    return mat;
}


/**
 * Write similarity values to an output file
 * @param output Output filename
 * @param mat Matrix of similarity values
 */
static void harry_write(char *output, hmatrix_t *mat)
{
    const char *cfg_str;

    /* Open output */
    config_lookup_string(&cfg, "output.output_format", &cfg_str);
    output_config(cfg_str);
    info_msg(1, "Writing %ld similarity values to '%0.40s' [%s].",
             mat->size, output, cfg_str);
    if (!output_open(output))
        fatal("Could not open output destination");

    output_write(mat);
    output_close();
}


/**
 * Exit Harry tool.
 */
static void harry_exit(hstring_t *strs, hmatrix_t *mat, int num)
{
    const char *cfg_str;

    /* Free memory */
    input_free(strs, num);
    free(strs);

    /* Destroy matrix */
    hmatrix_destroy(mat);

    config_lookup_string(&cfg, "input.stoptoken_file", &cfg_str);
    if (strlen(cfg_str) > 0)
        stoptokens_destroy();

    /* Destroy value cache */
    vcache_destroy();

    /* Destroy configuration */
    config_destroy(&cfg);
}

/**
 * Main function of Harry tool
 * @param argc Number of arguments
 * @param argv Argument values
 * @return exit code
 */
int main(int argc, char **argv)
{
    hmatrix_t *mat = NULL;
    hstring_t *strs = NULL;
    int num = 0;
    char *input = NULL, *input2 = NULL;
    char *output = NULL;

    harry_load_config(argc, argv);
    harry_parse_options(argc, argv, &input, &input2, &output);

    harry_init();
    strs = harry_read(input, input2, &num);
    mat = harry_alloc(strs, num);

    if (benchmark) {
        harry_benchmark(mat, strs, num);
    } else {
        harry_compute(mat, strs, num);
        harry_write(output, mat);
    }

    harry_exit(strs, mat, num);
    return EXIT_SUCCESS;
}
